app/controllers/sessions_controller.rb

```diff
@@ -5,16 +5,21 @@ def new
 
   def create
     user = User.find_by(email: params[:email])
-    session[:user_id] = user.id
-    if user.role == "admin"
-      flash[:success] = "Welcome Admin, #{user.name}!"
-      redirect_to(admin_dashboard_path)
-    elsif user.role == "merchant"
-      flash[:success] = "Welcome Merchant, #{user.name}!"
-      redirect_to(merchants_dashboard_path)
-    else
-      flash[:success] = "Welcome, #{user.name}. You are logged in!"
-      redirect_to(profile_path)
+    if user == nil || !user.authenticate(params[:password])
```

Comment:
Because we used `has_secure_password` in the user model, we have access to this nifty authenticate method (checks if password matches what we have in our db)!

Reply:
So before we were checking the roles of the user, with your new refactor we are checking if user is nil or if the user did not pass authentication via password, if that returns true, we get our first flash[:failure] (Can this be pulled from our ```user.errors``` that Brian showed us?). 

Our else condition is hit if the user exists OR if the user is authenticated via password (because the previous condition was an or, can this else statement get hit on accident if a user doesn't enter the correct password?). Once we hit the else, we see our session save and now we are differentiating by user.role

Reply:
When user enters invalid email, the user variable we have here returns `nil`. Because of this, when I tried to use the errors method that Brian showed us, I got a no method for nil class error. So this conditional is checking 1. is email inaccurate (aka user nil) *or* 2. is password incorrect.

Reply:
Got it! Thanks for the explanation!